### ISSUE TITLE
track! not converting

### ISSUE THREAD

**Issue by mikem836** [assaf/vanity#270]
I'm seeing the same problem as described in Issue #265. Everything works for the metrics, but conversions are not being recorded for ab tests. It works when using Vanity 1.9.2 but is broken in 2.0.0 beta.

Vanity (2.0.0.beta8)
Rails 4.2.3
Ruby 2.2.2

As described in #265, when I inspect the @hooks variable for the abtest, it's empty.

**Comment by jasonfb** [comment]
I was failing to do this for my first few attempts, and finally figured this out

you have to actually call the ab_test *before* the call to ```track!``` . if you fail to do that, then the 'current user' won't be in the experiment group and so can't be tracked

this has the unfortunate side effect of making the place in your app where you call ```track!``` have knowledge of all the experiments that contain that as a metric.

as a quick & dirty work-around to this brittleness, I devise this (somewhat hackish, but perfectly functional) solution, shown here associated with a 'completed order' metric (obviously this code goes alongside the completed order operation)  

```
          Vanity.playground.experiments.select{|k,v|
            v.metrics.collect(&:name).include?("Completed Order")
          }.keys.each do |vanity_experiment|
            ab_test(vanity_experiment)
          end
          track!(:completed_order)
```

**Comment by seanlinsley** [comment]
Just wanted to chime in, since we're also running into this issue.

We have Vanity integrated with Ahoy. We're using Ahoy's JS event tracking to also track events in ahoy. This is our code, with the fix:
```ruby
class Ahoy::Store < Ahoy::Stores::ActiveRecordStore
  # customize additional data to track per event
  def track_event(name, properties, options)
    super do |event|
      if /\Avanity\.(?<metric>.+)/ =~ name
        event.name = metric.to_s

        if ahoy.visitor_id
          # We are having issues with track! so we need the following lines.
          # Issue info: https://github.com/assaf/vanity/pull/271 https://github.com/assaf/vanity/issues/270
          Vanity.playground.metrics
          Vanity.playground.experiments

          Vanity.playground.track! metric.to_sym, identity: ahoy.visitor_id
        end
      end
    end
  end
end
```

**Comment by jasonfb** [comment]
do you actually call ab_test *before* you call ```track!``` (somewhere, anywhere, within the web request itself). 
If you fail to do that, your ```track!``` won't work

**Comment by jasonfb** [comment]
for our "add-to-cart" endpoint, we actually do this before we call ```track!```. This *programmatically* calls all of the experiments that are associated to the add-to-cart metric. 

```
Vanity.playground.experiments.select{ |k, v|
      v.metrics.collect(&:name).include?("Added To Cart")
    }.keys.each do |vanity_experiment|
      ab_test(vanity_experiment)
    end
```

**Comment by seanlinsley** [comment]
@jasonfb our metric tracking mostly happens as AJAX requests, with `ab_test` having been called while rendering templates for the page in question.

We can't actually call `ab_test` with our code because we exclude some users from any given AB test by doing something like this in the template:
```erb
<% if current_user && current_user.role == 'client' && ab_test(:some_test) == 'new_version' %>
  <!-- HTML for the new version -->
<% else %>
  <!-- HTML for the old version -->
<% end %>

**Comment by jasonfb** [comment]
Unless I'm mistaken Vanity won't work that way. 

All I'm saying is you have to call ```ab_test```  *again*, at least once, inside of the web request (ajax request) that is going to call ```track!```

We found that it was necessary to call ```ab_test``` before calling ```track!``` or else Vanity won't add that participant to the right group. 

In fact, it sounds like part of your problem is conceptual. You wrote: "We can't actually call ab_test with our code because we exclude some users from any given AB test" .... 

Consider this: Calling ab_test is what actually puts them into the group ( a or b, etc). Calling ```track!``` is what tracks them. Consider the first a prerequisite for the second. How can you call ```track!``` on someone you haven't yet put into a group? Doesn't make sense.

Presuming you want to restrict these to only users who have a role of 'client' (I am presuming that by reading your code), I see no reason why you couldn't use the example I wrote above (using ```.select``` to filter out only tests associated to the right metric), you could *still* wrap your calling code for ```ab_test``` AND ```track!``` in an ```if``` statement that checks for the special cases you want to check for.

**Comment by seanlinsley** [comment]
We shouldn't need to call `ab_test` because they've already been added to their buckets on initial page load, since `ab_test` is called in the template.

We already have the AB test knowledge duplicated in the Rails template and in the JS that sends the event. You're suggesting duplicating it again in our code that handles AJAX requests? Why would we do that when preloading `Vanity.playground.metrics` and `Vanity.playground.experiments` works just fine?

**Comment by phillbaker** [comment]
Sorry for the slow response on this, we'll get this fixed before a 2.0.0 release!

**Comment by seanlinsley** [comment]
Thanks for fixing this :smile_cat:
